File: cpp/demo-use-tools.cpp
// #define USE_IPEX
// #define USE_CUDA
// #define USE_BF16
// #define JITTER
#define OUTVID

#include <opencv2/opencv.hpp>
#include <opencv2/freetype.hpp>
#include <opencv2/videoio.hpp>
#include <torch/torch.h>
#include <torch/script.h>
#include <torch/nn/functional.h>
// #include <torchvision/vision.h>
// #include <argparse/argparse.hpp>
#include <random>
#include <chrono>
#include <iostream>
#include <fstream>
#include <filesystem>
#include "model.hpp"
#include "tools.hpp"
// #include "fail2link.hpp"

// #include<114/514.h>

#ifdef USE_IPEX
    #include <intel-extension-for-pytorch/csrc/cpu/vec/vec.h>
#endif
// argparse::ArgumentParser program("example-app");

// namespace fs = std::filesystem;

// program.add_argument("--video")
//     .help("input video path. live cam is used when not specified")
//     .default_value(std::string(""));
// program.add_argument("--model_weight")
//     .help("path to model weights file")
//     .default_value(std::string("data/model_weights.pth"));
// program.add_argument("--jitter")
//     .help("jitter bbox n times, and average results")
//     .default_value(0)
//     .scan<'i', int>();
// program.add_argument("-save_vis")
//     .help("saves output as video")
//     .default_value(false)
//     .implicit_value(true);
// program.add_argument("-save_text")
//     .help("saves output as text")
//     .default_value(false)
//     .implicit_value(true);
// program.add_argument("-display_off")
//     .help("do not display frames")
//     .default_value(false)
//     .implicit_value(true);
// program.add_argument("-cuda")
//     .help("use cuda")
//     .default_value(false)
//     .implicit_value(true);
// program.add_argument("-ipex")
//     .help("use intel extension for pytorch")
//     .default_value(false)
//     .implicit_value(true);
// program.add_argument("-bf16")
//     .help("use bfloat16")
//     .default_value(false)
//     .implicit_value(true);

// auto args = program.parse_args();

torch::Tensor preprocess_image(const cv::Mat& image) {
    // 调整图像大小
    cv::Mat resized_image;
    cv::resize(image, resized_image, cv::Size(224, 224));

    // 中心裁剪
    int crop_size = 224;
    int x = (resized_image.cols - crop_size) / 2;
    int y = (resized_image.rows - crop_size) / 2;
    cv::Rect crop_region(x, y, crop_size, crop_size);
    cv::Mat cropped_image = resized_image(crop_region);

    // 转换为张量
    torch::Tensor img_tensor = torch::from_blob(cropped_image.data, {cropped_image.rows, cropped_image.cols, 3}, torch::kByte);

    // 交换维度从HWC到CHW
    img_tensor = img_tensor.permute({2, 0, 1});

    // 转换为浮点型并归一化到[0, 1]
    img_tensor = img_tensor.toType(torch::kFloat) / 255.0;

    // 归一化
    img_tensor = torch::data::transforms::Normalize<>(
        {0.485, 0.456, 0.406}, 
        {0.229, 0.224, 0.225}
    )(img_tensor);

    return img_tensor;
}


// const std::string CNN_FACE_MODEL = "data/mmod_human_face_detector.dat"; // from http://dlib.net/files/mmod_human_face_detector.dat.bz2

#ifdef JITTER
std::tuple<float, float, float, float> bbox_jitter(float bbox_left, float bbox_top, float bbox_right, float bbox_bottom) {
    float cx = (bbox_right + bbox_left) / 2.0;
    float cy = (bbox_bottom + bbox_top) / 2.0;
    float scale = static_cast<float>(rand()) / RAND_MAX * 0.4 + 0.8;
    bbox_right = (bbox_right - cx) * scale + cx;
    bbox_left = (bbox_left - cx) * scale + cx;
    bbox_top = (bbox_top - cy) * scale + cy;
    bbox_bottom = (bbox_bottom - cy) * scale + cy;
    return std::make_tuple(bbox_left, bbox_top, bbox_right, bbox_bottom);
}
#endif
void drawrect(cv::Mat &image, cv::Rect rect, cv::Scalar color, int thickness) {
    cv::rectangle(image, rect, color, thickness);
}

void run(const std::string &video_path, const std::string &model_weight, int jitter, bool vis, bool display_off, bool save_text) {
    auto total_start_time = std::chrono::high_resolution_clock::now();

    // set up vis settings
    cv::Scalar colors[10];
    for (int i = 0; i < 10; ++i) {
        colors[i] = cv::Scalar(0, 255 * i / 10, 255 * (10 - i) / 10);
    }
    std::cerr<<"Start loading font file..."<<std::endl;
    cv::Ptr<cv::freetype::FreeType2> ft2;
    ft2 = cv::freetype::createFreeType2();
    ft2->loadFontData("data/arial.ttf", 0);
    std::cerr<<"Font file loaded."<<std::endl;

    // set up video source
    std::cerr<<"Start loading video..."<<std::endl;
    cv::VideoCapture cap;
    std::string video_output_path = video_path;
    if (video_path.empty()) {
        cap.open(0);
        video_output_path = "live.avi";
    } else {
        cap.open(video_path);
    }
    std::cerr<<"Video loaded."<<std::endl;

    // set up output file
    std::cerr<<"Start loading output file..."<<std::endl;
    std::ofstream f;
    if (save_text) {
        std::string outtext_name = std::filesystem::path(video_output_path).stem().string() + "_output.txt";
        f.open(outtext_name);
    }
    #ifdef OUTVID
    cv::VideoWriter outvid;
    if (vis) {
        std::string outvis_name = std::filesystem::path(video_output_path).stem().string() + "_output.avi";
        int imwidth = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_WIDTH));
        int imheight = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_HEIGHT));
        outvid.open(outvis_name, cv::VideoWriter::fourcc('M', 'J', 'P', 'G'), cap.get(cv::CAP_PROP_FPS), cv::Size(imwidth, imheight));
    }
    #endif
    if (!cap.isOpened()) {
        std::cerr << "Error opening video stream or file" << std::endl;
        return;
    }
    std::cerr<<"Output file loaded."<<std::endl;

    // set up data transformation
    // auto test_transforms = torch::data::transforms::Compose({
    //     torch::data::transforms::Resize(224),
    //     torch::data::transforms::CenterCrop(224),
    //     torch::data::transforms::ToTensor(),
    //     torch::data::transforms::Normalize({0.485, 0.456, 0.406}, {0.229, 0.224, 0.225})
    // });

    // load model weights
    std::cerr<<"Start loading model..."<<std::endl;
    ResNet model({3, 4, 6, 3}); // 使用您的层配置创建模型实例
    torch::load(model, model_weight); // 从保存的文件导入参数
    // torch::jit::script::Module model;
    // try {
    // // Deserialize the ScriptModule from a file using torch::jit::load().
    //     model = torch::jit::load("./data/traced_resnet_model.pt");
    // }
    // catch (const c10::Error& e) {
    //     std::cerr << "error loading the model\n";
    // }


    std::cerr<<"Model weights defined."<<std::endl;
    std::cerr<<"Model loaded."<<std::endl;
    #ifdef USE_CUDA
        model.to(torch::kCUDA);
    #endif
    model.eval();

    #ifdef USE_IPEX
        #ifdef USE_BF16
            model = ipex::optimize(model, torch::kBFloat16);
        #else
            model = ipex::optimize(model);
        #endif
    #endif
    std::cerr<<"Model set complete."<<std::endl;

    std::cerr<<"Start running reading loop..."<<std::endl;
    // video reading loop
    int frame_cnt = 0;
    while (cap.isOpened()) {
        auto frame_start_time = std::chrono::high_resolution_clock::now();
        cv::Mat frame;
        cap >> frame;
        if (frame.empty()) {
            break;
        }

        frame_cnt++;
        std::cerr<<"Getting face locations..."<<std::endl;
        auto bbox = get_face_loc(frame);
        std::cerr<<"Face locations got. Size:"<<bbox.size()<<std::endl<<"Start running face loop..."<<std::endl;
        if (!bbox.empty()) {
            for (const auto &b : bbox) {
                cv::Rect face_rect = b;
                auto face = frame(face_rect);
                auto img = preprocess_image(face);
                img.unsqueeze_(0);
                #ifdef JITTER
                if (jitter > 0) {
                    for (int i = 0; i < jitter; ++i) {
                        auto [bj_left, bj_top, bj_right, bj_bottom] = bbox_jitter(b[0], b[1], b[2], b[3]);
                        cv::Rect jittered_rect(bj_left, bj_top, bj_right - bj_left, bj_bottom - bj_top);
                        auto facej = frame(jittered_rect);
                        auto img_jittered = test_transforms(facej);
                        img_jittered.unsqueeze_(0);
                        img = torch::cat({img, img_jittered});
                    }
                }
                #endif
                // forward pass
                std::cerr<<"Start running forward pass..."<<std::endl;
                #ifdef USE_BF16
                    img = img.to(torch::kBFloat16);
                #endif
                torch::Tensor output;
                #ifdef USE_CUDA
                    output = model.forward(img.to(torch::kCUDA));
                #else
                    output = model.forward({img});
                #endif
                #ifdef JITTER
                if (jitter > 0) {
                    output = torch::mean(output, 0);
                }
                #endif
                float score = torch::sigmoid(output).item<float>();
                std::cerr<<"Forward pass done. Scode got."<<std::endl<<"Start drawing rect..."<<std::endl;

                int coloridx = std::min(static_cast<int>(std::round(score * 10)), 9);
                drawrect(frame, face_rect, colors[coloridx], 5);
                ft2->putText(frame, std::to_string(score), cv::Point(b.height, b.x), 40, cv::Scalar(255, 255, 255, 128), -1, cv::LINE_AA, true);
                std::cerr<<"Rect drawn."<<std::endl;
                if (save_text) {
                    f << frame_cnt << "," << score << "\n";
                }
            }
        }
        std::cerr<<"Face loop done."<<std::endl;
        if (!display_off) {
            cv::imshow("Result", frame);
            if (cv::waitKey(20) == 32) { // wait for space key to exit
                break;
            }
            #ifdef OUTVID
            if (vis) {
                outvid.write(frame);
            }
            #endif
        }

        auto frame_end_time = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> frame_time = frame_end_time - frame_start_time;
        double frame_rate = 1.0 / frame_time.count();
        std::cout << "frame time: " << frame_time.count() << "\trealtime frame rate: " << frame_rate << std::endl;
    }
    #ifdef OUTVID
    if (vis) {
        outvid.release();
    }
    #endif
    if (save_text) {
        f.close();
    }
    cap.release();
    std::cout << "DONE!" << std::endl;

    auto total_end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> total_time = total_end_time - total_start_time;
    std::cout << "total time: " << total_time.count() << std::endl;
}

int main(int argc, char **argv) {
    // program.parse_args(argc, argv);
    // run(args.get<std::string>("video"), args.get<std::string>("model_weight"), args.get<int>("jitter"), args.get<bool>("save_vis"), args.get<bool>("display_off"), args.get<bool>("save_text"), args.get<bool>("cuda"), args.get<bool>("ipex"), args.get<bool>("bf16"));
    run("./videoplayback.avi","./data/model_weights.pth",0,false,true,false);
    return 0;
}

File: eyecontactcnn/demo-use-tools.py
# import dlib
import cv2
import argparse, os, random
import torch
import torch.nn as nn
import torch.nn.functional as F
import torchvision
from torchvision import datasets, transforms
import pandas as pd
import numpy as np
from model import model_static
from PIL import Image
from PIL import ImageDraw
from PIL import ImageFont
from colour import Color
from tools import get_face_loc
import time 

import intel_extension_for_pytorch as ipex


parser = argparse.ArgumentParser()

parser.add_argument('--video', type=str, help='input video path. live cam is used when not specified')
parser.add_argument('--model_weight', type=str, help='path to model weights file', default='data/model_weights.pkl')
parser.add_argument('--jitter', type=int, help='jitter bbox n times, and average results', default=0)
parser.add_argument('-save_vis', help='saves output as video', action='store_true')
parser.add_argument('-save_text', help='saves output as text', action='store_true')
parser.add_argument('-display_off', help='do not display frames', action='store_true')
parser.add_argument('-cuda', help='use cuda', action='store_true')
parser.add_argument('-ipex', help='use intel extension for pytorch', action='store_true')
parser.add_argument('-bf16', help='use bfloat16', action='store_true')

args = parser.parse_args()

CNN_FACE_MODEL = 'data/mmod_human_face_detector.dat' # from http://dlib.net/files/mmod_human_face_detector.dat.bz2


def bbox_jitter(bbox_left, bbox_top, bbox_right, bbox_bottom):
    cx = (bbox_right+bbox_left)/2.0
    cy = (bbox_bottom+bbox_top)/2.0
    scale = random.uniform(0.8, 1.2)
    bbox_right = (bbox_right-cx)*scale + cx
    bbox_left = (bbox_left-cx)*scale + cx
    bbox_top = (bbox_top-cy)*scale + cy
    bbox_bottom = (bbox_bottom-cy)*scale + cy
    return bbox_left, bbox_top, bbox_right, bbox_bottom


def drawrect(drawcontext, xy, outline=None, width=0):
    (x1, y1), (x2, y2) = xy
    points = (x1, y1), (x2, y1), (x2, y2), (x1, y2), (x1, y1)
    drawcontext.line(points, fill=outline, width=width)


def run(video_path, model_weight, jitter, vis, display_off, save_text, USE_CUDA=False, USE_IPEX=False, USE_BF16=False):
    total_start_time = time.time()
    # set up vis settings
    red = Color("red")
    colors = list(red.range_to(Color("green"),10))
    font = ImageFont.truetype("data/arial.ttf", 40)

    # set up video source
    if video_path is None:
        cap = cv2.VideoCapture(0)
        video_path = 'live.avi'
    else:
        cap = cv2.VideoCapture(video_path)

    # set up output file
    if save_text:
        outtext_name = os.path.basename(video_path).replace('.avi','_output.txt')
        f = open(outtext_name, "w")
    if vis:
        outvis_name = os.path.basename(video_path).replace('.avi','_output.avi')
        imwidth = int(cap.get(3)); imheight = int(cap.get(4))
        outvid = cv2.VideoWriter(outvis_name,cv2.VideoWriter_fourcc('M','J','P','G'), cap.get(5), (imwidth,imheight))

    # set up face detection mode
    # if face_path is None:
    #     facemode = 'DLIB'
    # else:
    #     facemode = 'GIVEN'
    #     column_names = ['frame', 'left', 'top', 'right', 'bottom']
    #     df = pd.read_csv(face_path, names=column_names, index_col=0)
    #     df['left'] -= (df['right']-df['left'])*0.2
    #     df['right'] += (df['right']-df['left'])*0.2
    #     df['top'] -= (df['bottom']-df['top'])*0.1
    #     df['bottom'] += (df['bottom']-df['top'])*0.1
    #     df['left'] = df['left'].astype('int')
    #     df['top'] = df['top'].astype('int')
    #     df['right'] = df['right'].astype('int')
    #     df['bottom'] = df['bottom'].astype('int')

    if (cap.isOpened()== False):
        print("Error opening video stream or file")
        exit()

    # if facemode == 'DLIB':
    #     cnn_face_detector = dlib.cnn_face_detection_model_v1(CNN_FACE_MODEL)
    frame_cnt = 0

    # set up data transformation
    test_transforms = transforms.Compose([transforms.Resize(224), transforms.CenterCrop(224), transforms.ToTensor(),
                                         transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])])

    # load model weights
    model = model_static(model_weight,USE_CUDA=USE_CUDA)
    model_dict = model.state_dict()
    torch.set_num_threads(2*torch.get_num_threads())
    if USE_CUDA:
        snapshot = torch.load(model_weight)
    else:
        snapshot = torch.load(model_weight, map_location=torch.device('cpu'))
    model_dict.update(snapshot)
    model.load_state_dict(model_dict)

    if USE_CUDA:
        model.cuda()
    model.train(False)

    if USE_IPEX:
        if USE_BF16:
            model = ipex.optimize(model=model,dtype=torch.bfloat16)
        else:
            model = ipex.optimize(model=model)
    # video reading loop
    while(cap.isOpened()):
        frame_start_time = time.time()
        ret, frame = cap.read()
        print("readed")
        if ret == True:
            height, width, channels = frame.shape
            frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)

            frame_cnt += 1
            bbox = get_face_loc(frame)
            frame = Image.fromarray(frame)
            if bbox is not []:
                for b in bbox:
                    face = frame.crop((b))
                    img = test_transforms(face)
                    img.unsqueeze_(0)
                    if jitter > 0:
                        for i in range(jitter):
                            bj_left, bj_top, bj_right, bj_bottom = bbox_jitter(b[0], b[1], b[2], b[3])
                            bj = [bj_left, bj_top, bj_right, bj_bottom]
                            facej = frame.crop((bj))
                            img_jittered = test_transforms(facej)
                            img_jittered.unsqueeze_(0)
                            img = torch.cat([img, img_jittered])

                    # forward pass
                    if USE_BF16:
                        img = img.bfloat16()
                    if USE_CUDA:
                        output = model(img.cuda())
                    else:
                        output = model(img)
                    if jitter > 0:
                        output = torch.mean(output, 0)
                    score = F.sigmoid(output).item()

                    coloridx = min(int(round(score*10)),9)
                    draw = ImageDraw.Draw(frame)
                    drawrect(draw, [(b[0], b[1]), (b[2], b[3])], outline=colors[coloridx].hex, width=5)
                    draw.text((b[0],b[3]), str(round(score,2)), fill=(255,255,255,128), font=font)
                    if save_text:
                        f.write("%d,%f\n"%(frame_cnt,score))

            if not display_off:
                frame = np.asarray(frame) # convert PIL image back to opencv format for faster display
                frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
                cv2.imshow('Result',frame)
                if cv2.waitKey()  == ord(' '):       # 每帧图像显示20ms，等待用户按下空格键退出
                    break
                print("showing")
                if vis:
                    outvid.write(frame)
        else:
            break
        frame_end_time = time.time()
        frame_time = frame_end_time - frame_start_time
        frame_rate = 1/frame_time
        print("frame time:",frame_time,"\trealtime frame rate:",frame_rate)

    if vis:
        outvid.release()
    if save_text:
        f.close()
    cap.release()
    print('DONE!')
    total_end_time = time.time()
    total_time = total_end_time - total_start_time
    print("total time:",total_time)


if __name__ == "__main__":
    run(args.video, args.model_weight, args.jitter, args.save_vis, args.display_off, args.save_text, args.cuda, args.ipex, args.bf16)